security/pkg/nodeagent/caclient/providers/google-cas/client.go

```diff
@@ -0,0 +1,145 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package caclient
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	privateca "cloud.google.com/go/security/privateca/apiv1"
+	"google.golang.org/api/option"
+	privatecapb "google.golang.org/genproto/googleapis/cloud/security/privateca/v1"
+	"google.golang.org/protobuf/types/known/durationpb"
+	"k8s.io/apimachinery/pkg/util/rand"
+
+	"istio.io/istio/pkg/security"
+	"istio.io/pkg/log"
+)
+
+var googleCAClientLog = log.RegisterScope("googlecas", "Google CAS client debugging", 0)
+
+// GoogleCASClient: Agent side plugin for Google CAS
+type GoogleCASClient struct {
+	caSigner string
+	caClient *privateca.CertificateAuthorityClient
+}
+
+// NewGoogleCASClient create a CA client for Google CAS.
+func NewGoogleCASClient(capool string, options ...option.ClientOption) (security.Client, error) {
+	caClient := &GoogleCASClient{caSigner: capool}
+	ctx := context.Background()
+	var err error
+
+	caClient.caClient, err = privateca.NewCertificateAuthorityClient(ctx, options...)
+
+	if err != nil {
+		googleCAClientLog.Errorf("unable to initialize google cas caclient: %v", err)
+		return nil, err
+	}
+	googleCAClientLog.Debugf("Intitialized Google CAS plugin with endpoint: %v", capool)
+	return caClient, nil
+}
+
+func (r *GoogleCASClient) createCertReq(name string, csrPEM []byte, lifetime time.Duration) *privatecapb.CreateCertificateRequest {
+	var isCA bool = false
+
+	// We use Certificate_Config option to ensure that we only request a certificate with CAS supported extensions/usages.
+	// CAS uses the PEM encoded CSR only for its public key and infers the certificate SAN (identity) of the workload through SPIFFE identity reflection
+	creq := &privatecapb.CreateCertificateRequest{
+		Parent:        r.caSigner,
+		CertificateId: name,
+		Certificate: &privatecapb.Certificate{
+			Lifetime: durationpb.New(lifetime),
+			CertificateConfig: &privatecapb.Certificate_Config{
+				Config: &privatecapb.CertificateConfig{
+					X509Config: &privatecapb.X509Parameters{
+						KeyUsage: &privatecapb.KeyUsage{
+							BaseKeyUsage: &privatecapb.KeyUsage_KeyUsageOptions{
+								DigitalSignature: true,
+								KeyEncipherment:  true,
+							},
+							ExtendedKeyUsage: &privatecapb.KeyUsage_ExtendedKeyUsageOptions{
+								ServerAuth: true,
+								ClientAuth: true,
+							},
+						},
+						CaOptions: &privatecapb.X509Parameters_CaOptions{
+							IsCa: &isCA,
+						},
+					},
+					PublicKey: &privatecapb.PublicKey{
+						Format: privatecapb.PublicKey_PEM,
+						Key:    csrPEM,
+					},
+				},
+			},
+			SubjectMode: privatecapb.SubjectRequestMode_REFLECTED_SPIFFE,
+		},
+	}
+	return creq
+}
+
+// CSR Sign calls Google CAS to sign a CSR.
+func (r *GoogleCASClient) CSRSign(csrPEM []byte, certValidTTLInSec int64) ([]string, error) {
+	certChain := []string{}
+	name := fmt.Sprintf("csr-workload-%s", rand.String(8))
```

Comment:
Do you need to call Seed first?
https://pkg.go.dev/k8s.io/apimachinery/pkg/util/rand#Seed

Reply:
done